velox/exec/tests/PrefixSortTest.cpp

```diff
@@ -0,0 +1,355 @@
+using namespace facebook::velox::test;
+
+namespace facebook::velox::exec::prefixsort::test {
+
+class TestSetting {
+ public:
+  TestSetting(
+      const std::vector<CompareFlags>& sortCompareFlags,
+      const RowVectorPtr& expectedResult)
+      : sortCompareFlags_(sortCompareFlags), expectedResult_(expectedResult){};
+
+  const std::vector<CompareFlags>& sortCompareFlags() const {
+    return sortCompareFlags_;
+  }
+
+  const RowVectorPtr& expectedResult() const {
+    return expectedResult_;
+  }
+
+  std::string debugString() const {
+    const auto numRows = expectedResult_->size();
+    const std::string expectedResultStr =
+        expectedResult_->toString(0, numRows > 100 ? 100 : numRows);
+    std::stringstream sortCompareFlagsStr;
+    for (const auto sortCompareFlag : sortCompareFlags_) {
+      sortCompareFlagsStr << sortCompareFlag.toString();
+    }
+    return fmt::format(
+        "sortCompareFlags:{}, \nexpectedResult:\n{}",
+        sortCompareFlagsStr.str(),
+        expectedResultStr);
+  }
+
+ private:
+  std::vector<CompareFlags> sortCompareFlags_;
+  RowVectorPtr expectedResult_;
+};
+
+class PrefixSortTest : public OperatorTestBase {
+ protected:
+  std::vector<char*>
+  storeRows(int numRows, const RowVectorPtr& sortedRows, RowContainer* data);
+
+  void testSort(
+      int numRows,
+      const std::vector<TestSetting> testSettings,
+      const RowVectorPtr& sortedRows);
+
+  // Use std::sort to generate expected result.
+  const RowVectorPtr generateExpectedResult(
+      const std::vector<CompareFlags> compareFlags,
+      int numRows,
+      const RowVectorPtr& sortedRows);
+
+  // 20 MB
+  const int64_t maxBytes_ = 20LL << 20;
+  const std::shared_ptr<memory::MemoryPool> rootPool_{
+      memory::MemoryManager::getInstance()->addRootPool(
+          "PrefixSortTest",
+          maxBytes_)};
+
+  const std::shared_ptr<memory::MemoryPool> pool_{
+      rootPool_->addLeafChild("PrefixSortTest", maxBytes_)};
+
+  const std::shared_ptr<memory::MemoryPool> fuzzerPool =
+      rootPool_->addLeafChild("VectorFuzzer");
+};
+
+std::vector<char*> PrefixSortTest::storeRows(
+    int numRows,
+    const RowVectorPtr& sortedRows,
+    RowContainer* data) {
+  std::vector<char*> rows;
+  SelectivityVector allRows(numRows);
+  rows.resize(numRows);
+  for (int row = 0; row < numRows; ++row) {
+    rows[row] = data->newRow();
+  }
+  for (int column = 0; column < sortedRows->childrenSize(); ++column) {
+    DecodedVector decoded(*sortedRows->childAt(column), allRows);
+    for (int i = 0; i < numRows; ++i) {
+      char* row = rows[i];
+      data->store(decoded, i, row, column);
+    }
+  }
+  return rows;
+}
+
+const RowVectorPtr PrefixSortTest::generateExpectedResult(
+    const std::vector<CompareFlags> compareFlags,
+    int numRows,
+    const RowVectorPtr& sortedRows) {
+  const auto sortedRowsType = asRowType(sortedRows->type());
+  const int sortKeysNum = sortedRows->childrenSize();
```

Comment:
This is wrong. It should be compareFlags.size(). This is failing in fuzzMulti test.

Reply:
e , yes , it`s wrong , but do not affect test result because when sortKeys are equal than compare payload, but I only need to assert that keys are equal....